Updated box and item checks to allow multiple barcodes separated by a…

westbl opened the pull request:
Updated box and item checks to allow multiple barcodes separated by a semicolon as a delimiter. Currently using a single string as storage; "better" methods (allowing multiple barcodes in the database) would require heavy database updates.

This update should ensure that box and item barcode(s) are unique. If implemented, the database should be updated to merge duplicate boxes and items (probably a manual procedure).

westbl committed:
Updated box and item checks to allow multiple barcodes separated by a semicolon as a delimiter

westbl committed:
Updated box and item checks to allow multiple barcodes separated by a semicolon as a delimiter

bradjc commented on chezbetty/models/box.py:
```diff
@@ -27,7 +28,25 @@ def from_id(cls, id):
     def from_barcode(cls, barcode):
-        return DBSession.query(cls).filter(cls.barcode == barcode).one()
+        box_list = DBSession.query(cls).filter(cls.barcode.ilike('%{}%'.format(barcode))).all()
+        if(len(box_list) == 1):
```
```suggestion
        if len(box_list) == 1:
```

bradjc commented on chezbetty/models/box.py:
```diff
@@ -27,7 +28,25 @@ def from_id(cls, id):
+        #TODO: handle the two cases where len(box_list) != 1 (len is 0, len is > 1)
+
+    @classmethod #search for all delimited barcodes in database; return true if any matches are found
+    def update_exists_barcode(cls, barcode, id):
```
Should this be `barcode_exists()`? It doesn't update does it?

westbl replied:
So this is what I called the function which is used when updating a box or item in the database, rather than creating a new item. The ID is required so that it doesn't falsely detect itself as a conflict, which is why there are separate functions. The naming might be odd, but I wasn't sure what else to call it...it is checking if the barcode exists in the admin_box_edit_submit and admin_edit_item_submit functions.

bradjc commented on chezbetty/models/box.py:
```diff
@@ -27,7 +28,25 @@ def from_id(cls, id):
+                result = DBSession.query(func.count(cls.id).label('c'))\
+                            .filter(cls.id != id)\
+                            .filter(cls.barcode.ilike('%{}%'.format(single_barcode))).one().c > 0
+                if(result):
```
```suggestion
                if result:
```

bradjc commented on chezbetty/models/box.py:
```diff
@@ -27,7 +28,25 @@ def from_id(cls, id):
+        sub = re.compile(r'[^\d;]+').sub('', barcode).split(';') #remove any characters that aren't digits or delimiters
+        result = False
+        for single_barcode in sub:
+            if single_barcode is not "":
```
```suggestion
            if single_barcode != "":
```

bradjc commented on chezbetty/models/item.py:
```diff
@@ -102,20 +103,29 @@ def exists_name(cls, name):
+        sub = re.compile(r'[^\d;]+').sub('', barcode).split(';') #remove any characters that aren't digits or delimiters
+        result = False
+        for single_barcode in sub:
+            if single_barcode is not "":
```
```suggestion
            if single_barcode != "":
```

bradjc commented on chezbetty/models/item.py:
```diff
@@ -102,20 +103,29 @@ def exists_name(cls, name):
+            if single_barcode is not "":
+                result = DBSession.query(func.count(cls.id).label('c'))\
+                            .filter(cls.barcode.ilike('%{}%'.format(single_barcode))).one().c > 0
+                if(result):
```
```suggestion
                if result:
```

bradjc reviewed (commented):
I don't understand the need for the `update_` functions.

Also, I wonder if this will have any effect on purchasing performance because of the slower item lookups. It may not matter in practice.

westbl committed:
Merged exists_barcode and update_exists_barcode functions; changed styling

westbl committed:
Merged exists_barcode and update_exists_barcode functions; changed styling

westbl commented:
@bradjc I updated the code to merge the "update_exists_barcode" and "exists_barcode" functions by giving the "id" variable a default value of "None".  I am not sure if it will break the query (I don't know what value Python will put as None in the equality check), but it seemed to work when I tested it locally.

I also updated the code style to match your other comments.  Do you like this better?

westbl committed:
Added instruction prompt for users with <= 3 purchases and 0 all-time deposits

westbl committed:
Merge branch 'multi-barcode' of https://github.com/cseg-michigan/chez-betty into multi-barcode

bradjc commented on chezbetty/models/item.py:
```diff
@@ -102,20 +98,17 @@ def exists_name(cls, name):
-                        .filter(cls.barcode.ilike('%{}%'.format(single_barcode))).one().c > 0
-            if(result):
-                return result
+            if single_barcode is not "":
```
```suggestion
            if single_barcode != "":
```

bradjc reviewed (commented):
Yeah looks good.

westbl committed:
Fixed code style and merged with changes from Master branch
